keps/369-job-interface/README.md

```diff
@@ -0,0 +1,423 @@
+know that this has succeeded?
+-->
+
+- Define an interface which shapes the default behaviors of Job
+- Make Kubernetes Job an implementation template of the interface
+- Provide a library to organize helper functions which is used in constructing controllers
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Integrate with other job-like applications
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+We collected feedbacks from the community about how to fully integrate Kueue with MPIJob,
+see [#499](https://github.com/kubernetes-sigs/kueue/issues/499).
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Job interface defined here is a hint for developers to build their own controllers,
+it's not a hard constraint(to implement the interface) but a good reference.
+- This will increase the code complexity by wrapping the original Jobs.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Part I: Job Interface
+
+We will define a new interface named GenericJob, this should be implemented by custom job-like applications:
+
+```golang
+type GenericJob interface {
+  // Object returns the job instance.
+  Object() client.Object
+  // IsSuspend returns whether the job is suspend or not.
+  IsSuspend() bool
+  // Suspend will suspend the job.
+  Suspend() error
+  // UnSuspend will unsuspend the job.
+  UnSuspend() error
```

Comment:
How will the library/controller know which job(and of what kind) to unsuspend because a particular workload got admitted? There is no key/identrifier returned by this Interface.